clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1930,24 +1955,35 @@ class SyclKernelPointerHandler : public SyclKernelFieldHandler {
     return ModifiedRD;
   }
 
+  bool isArrayElement(const FieldDecl *FD, QualType Ty) const {
+    return !SemaRef.getASTContext().hasSameType(FD->getType(), Ty);
+  }
+
 public:
   static constexpr const bool VisitInsideSimpleContainersWithPointer = true;
+  static constexpr const bool VisitNthArrayElement = false;
   SyclKernelPointerHandler(Sema &S, const CXXRecordDecl *RD)
       : SyclKernelFieldHandler(S) {
     createNewType(RD);
   }
 
+  SyclKernelPointerHandler(Sema &S) : SyclKernelFieldHandler(S) {}
+
   bool enterStruct(const CXXRecordDecl *, FieldDecl *, QualType Ty) final {
     createNewType(Ty->getAsCXXRecordDecl());
     return true;
   }
 
   bool leaveStruct(const CXXRecordDecl *, FieldDecl *FD, QualType Ty) final {
+
     CXXRecordDecl *ModifiedRD = getGeneratedNewRecord(Ty->getAsCXXRecordDecl());
 
-    // Add this record as a field of it's parent record.
-    if (!ModifiedRecords.empty())
+    if (!isArrayElement(FD, Ty))
```

Comment:
Could you please add a comment here? Thanks

Reply:
Sorry can you elaborate on what you mean by handle multi-dimensional array element? This code prevents adding array elements as a field of it's parent record. When we leave a struct, we check if the struct is a field or array element. If its a field, we add it to it's parent record. If not, we don't add it (because the field be be array of this struct and not the struct itself.